### PR TITLE
rustc_metadata: use a table for super_predicates, fn_sig, impl_trait_ref.

### PR DESCRIPTION
**Pull request by eddyb**
This is an attempt at a part of #65407, i.e. moving parts of cross-crate "metadata" into tables that match queries more closely.

Three new tables should be enough to see some perf/metadata size changes.
(need to do something similar to https://github.com/rust-lang/rust/pull/59953#issuecomment-542521919)

There are other bits of data that could be made into tables, but they can be more compact so the impact would likely be not as bad, and they're also more work to set up.

### CODE REVIEW

**Comment by eddyb** [comment]
@rust-timer build 238668d24087abeafc989aadc78ee7c41a3c59d6

**Comment by eddyb** [comment]
This is smaller increase than https://github.com/rust-lang/rust/pull/59953#issuecomment-542521919, which is a good sign:

| | Before (518deda77feb4957bfd311b6cb50baa7ef9ca6a2) <br> [rust-std.tar.xz](https://rust-lang-ci2.s3-us-west-1.amazonaws.com/rustc-builds/518deda77feb4957bfd311b6cb50baa7ef9ca6a2/rust-std-nightly-x86_64-unknown-linux-gnu.tar.xz) | After (238668d24087abeafc989aadc78ee7c41a3c59d6) <br> [rust-std.tar.xz](https://rust-lang-ci2.s3-us-west-1.amazonaws.com/rustc-builds/238668d24087abeafc989aadc78ee7c41a3c59d6/rust-std-nightly-x86_64-unknown-linux-gnu.tar.xz) | Increase |
| --- | --: | --: | --- |
| `libcore.rlib` | 25.85MB | 26.12MB | 1.0% |
| `libstd.rlib` | 10.12MB | 10.23MB | 1.1% |
| Total | 178.36MB | 178.60MB | 0.1% |

(Note that the total is compressed with XZ, but also includes `librustc*`, soon to be in a separate `rustc-dev` component)

File: src/librustc_metadata/encoder.rs [hunk 1]
```diff
@@ -1321,10 +1327,11 @@ impl EncodeContext<'tcx> {
     fn encode_info_for_closure(&mut self, def_id: DefId) {
         debug!("EncodeContext::encode_info_for_closure({:?})", def_id);
 
-        let tables = self.tcx.typeck_tables_of(def_id);
+        // FIXME(eddyb) would `tcx.type_of(def_id)` work as well?
```
Comment (michaelwoerister) [inline src/librustc_metadata/encoder.rs]: Can you try that? It looks like it should.

File: src/librustc_metadata/encoder.rs [hunk 2]
```diff
@@ -635,13 +641,8 @@ impl EncodeContext<'tcx> {
         let data = VariantData {
             ctor_kind: variant.ctor_kind,
             discr: variant.discr,
-            // FIXME(eddyb) deduplicate these with `encode_enum_variant_ctor`.
+            // FIXME(eddyb) deduplicate this with `encode_enum_variant_ctor`.
```
Comment (michaelwoerister) [inline src/librustc_metadata/encoder.rs]: I don't understand what "duplicate this with" means. Do you mean "keep this in sync with"?

File: src/librustc_metadata/encoder.rs [hunk 3]
```diff
@@ -660,6 +661,10 @@ impl EncodeContext<'tcx> {
         self.encode_deprecation(def_id);
         self.encode_item_type(def_id);
         if variant.ctor_kind == CtorKind::Fn {
+            if let Some(ctor_def_id) = variant.ctor_def_id {
+                // FIXME(eddyb) deduplicate this with `encode_enum_variant_ctor`.
```
Comment (michaelwoerister) [inline src/librustc_metadata/encoder.rs]: Same as above.

**Review by michaelwoerister — COMMENTED** [review]
Looks like an improvement! r=me with the nits addressed.

File: src/librustc_metadata/encoder.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_metadata/encoder.rs]: "deduplicate" as in there should be a helper or (even better) this code should exist only in one place.
The comment is old anyway, I think it predates @petrochenkov's revamp of how enum/struct ctors work.

File: src/librustc_metadata/encoder.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_metadata/encoder.rs]: I suppose it's easy enough to check.

File: src/librustc_metadata/encoder.rs [hunk 4]
```diff
@@ -1321,10 +1328,12 @@ impl EncodeContext<'tcx> {
     fn encode_info_for_closure(&mut self, def_id: DefId) {
         debug!("EncodeContext::encode_info_for_closure({:?})", def_id);
 
-        let tables = self.tcx.typeck_tables_of(def_id);
+        // NOTE(eddyb) `tcx.type_of(def_id)` isn't used because it's fully generic,
+        // including on the signature, which is inferred in `typeck_tables_of.
```
Comment (michaelwoerister) [inline src/librustc_metadata/encoder.rs]: Thanks for finding that out! Makes sense.

**Pull request merged**